### PR TITLE
Components: Add filter picker

### PR DESCRIPTION
**Pull request by psealock**
This PR adds the Filter Picker to a Products Report. The filter picker accepts a configuration array and is capable of supporting multiple levels of subFilters.

**Designs** p6riRB-3lk-p2

### For the next PRs
* Animations
* Search component for selecting a single product

### Screeshots
Closed
screen shot 2018-07-16 at 4 17 34 pm

Open
screen shot 2018-07-16 at 4 17 43 pm

Search Component (to be created in following PR)
screen shot 2018-07-16 at 4 18 02 pm

### Testing

1. Go to `/wp-admin/admin.php?page=wc-admin#/analytics/products`
2. Make selections and see the url parameter update (except "Single Product")
3. Add other parameters by using the datepicker and ensure the Filter Picker doesn't interfere with those
4. Make sure accessibility via keyboard works as expected
5. Make sure the selection is correctly referenced in the UI. "All Products" is the default
6. Compare the styles (selected, hover, focus) with datepicker selection and note any differences
7. Test different screen widths to make sure the Datepicker and Filterpicker are rendered correctly

Fixes https://github.com/woocommerce/wc-admin/issues/191

### CODE REVIEW

File: client/components/date-picker/README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-Date Picker (Work in Progress)
+Date Picker
```
Comment (timmyc) [inline client/components/date-picker/README.md]: 🎉

File: client/components/filter-picker/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+const renderFilterPicker = ( { path, query } ) {
+	const filters = [
+		{ label: 'Breakfast', value: 'breakfast' },
+		{ label: 'Lunch', value: 'lunch', subFilters: [
+			{ label: 'Meat', value: 'meat' },
+			{ label: 'Vegan', value: 'vegan' },
+			{ label: 'Pescatarian', value: 'fish', subFilters: [
+				{ label: 'Snapper', value: 'snapper' },
+				{ label: 'Cod', value: 'cod' },
```
Comment (timmyc) [inline client/components/filter-picker/README.md]: I'm fond of salmon myself.

File: client/components/filter-picker/index.js [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+	getSelectedFilter() {
+		const { filters, getQueryParamValue, filterPaths } = this.props;
+		const value = getQueryParamValue();
+		const filterPath = filterPaths[ value ];
+		const visibleFilters = this.getVisibleFilters( filters, [ ...filterPath ] );
+		return find( visibleFilters, filter => filter.value === value );
+	}
+
+	getLabels( selectedFilter ) {
+		return [ selectedFilter && selectedFilter.label ];
```
Comment (timmyc) [inline client/components/filter-picker/index.js]: If `selectedFilter` is null/falsey couldn't this result in a `[ null ]` being returned here?

File: client/components/filter-picker/index.js [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		const nav = [ ...this.state.nav ];
+		nav.pop();
+		this.setState( { nav } );
+		this.focusFirstFilter();
+	}
+
+	focusFirstFilter() {
+		setTimeout( () => {
+			const list = this.listRef.current;
+			list.children[ 0 ].children[ 0 ].focus();
```
Comment (timmyc) [inline client/components/filter-picker/index.js]: Do we need null checks on the children attributes here?

File: client/components/filter-picker/index.js [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+	goBack() {
+		const nav = [ ...this.state.nav ];
+		nav.pop();
+		this.setState( { nav } );
+		this.focusFirstFilter();
+	}
+
+	focusFirstFilter() {
+		setTimeout( () => {
```
Comment (timmyc) [inline client/components/filter-picker/index.js]: why is setTimeout needed here?

**Review by timmyc — COMMENTED** [review]
Working nicely for me. Tab navigation through the dropdown is working well, as are small viewports, and retention of selections on both the Date Picker and the Filter picker when hard refreshing the page.

Just a couple of quick comments, otherwise looking dandy!

File: client/components/filter-picker/index.js (hunk 5 above)
Comment (psealock) [reply client/components/filter-picker/index.js]: To ensure `focus()` gets called at the end of the call stack when the elements are rendered.

**Review by timmyc — APPROVED** [review]
Thanks for the updates LGTM

**Pull request merged**